content/05_Filtering.md

```diff
@@ -1,505 +1,427 @@
+
+If Rx were being designed from scratch today, the relevant operators in the preceding section would just have the normal names: `First`, and `FirstOrDefault`, and so on. The reason they all end with `Async` is that these were added in Rx 2.0, and Rx 1.0 had already defined operators with those names. This example uses the `First` operator:
+
+```cs
+int v = Observable.Range(1, 10).First();
+Console.WriteLine(v);
+```
+
+This prints out the value `1`, which is the first item returned by `Range` here. But look at the type of that variable `v`. It's not an `IObservable<int>`, it's just an `int`. What would happen if we used this on an Rx operator that didn't immediately produce values upon subscription? Here's one example:
+
+```cs
+long v = Observable.Timer(TimeSpan.FromSeconds(2)).First();
+Console.WriteLine(v);
+```
+
+If you run this, you'll find that the call to `First` doesn't return until a value is produced. It is a _blocking_ operator. We typically avoid blocking operators in Rx, because it's easy to create deadlocks with them. The whole point of Rx is that we can create code that reacts to events, so to just sit and wait until a specific observable source produces a value is not really in the spirit of things. If you find yourself wanting to do that, there are often better ways to achieve the results you're looking for. (Or perhaps Rx isn't good model for whatever you're doing.)
+
+If you really do need to wait for a value like this, it might be better to use the `Async` forms in conjunction with Rx's integrated support for C#'s `async`/`await`:
+
+```cs
+long v = await Observable.Timer(TimeSpan.FromSeconds(2)).FirstAsync();
+Console.WriteLine(v);
+```
+
+This logically has the same effect, but because we're using `await`, this won't block the calling thread while it waits for the observable source to produce a value. This might reduce the chances of deadlock.
+
+The fact that we're able to use `await` makes some sense of the fact that these methods end with `Async`, but you might be wondering what's going on here. We've seen that these methods all return `IObservable<T>`, not `Task<T>`, so how are we able to use `await`? There's a full explanation in the [Leaving Rx's World](13_LeavingIObservable.md) chapter, but the short answer is that Rx provides extension methods that enable this to work. When you `await` an observable sequence, the `await` will complete once the source completes, and it will return the final value that emerges from the source. This works well for operators such as `FirstAsync` and `LastAsync` that produce exactly one item.
```

Comment:
As it says, there is a full explanation in a different chapter. Since this chapter is about filtering, it seemed to make more sense to put the description of how this works in the chapter that is already dedicated to integration between Rx and non-Rx stuff.

I've modified the link to point to the relevant section of the chapter, but I think that's enough—I don't think it would be an improvement to move that content out of that chapter and into this one.

> It would be quite nice to have a bit more technical detail about how this works - is it using the awaiter approach where you can make anything awaitable (like how foreach works)?

I'm not sure what you mean, because `await foreach` uses a different mechanism from all other `await` expressions. I _think_ (but I'm not sure) maybe you're referring to the pattern-based approach: `GetAwaiter` for `await` is a pattern just like `GetEnumerator` is a pattern for `foreach`, and all of the LINQ query expression stuff also expects similar well-known method names.

If that's what you meant, then yes. I don't want to put that in a chapter on filtering, because we're already one digression deep by talking about `await`, and this would takes us further away. I think it could go in the part where I discuss how `await` integration works, in Ch13's "Integration with `async` and `await`" section, but even there, my gut feeling is that drawing a comparison with `foreach` would be more distracting than illuminating.